### PR TITLE
[WIP] Migrate ntp si

### PR DESCRIPTION
**Pull request by deeppandya**
Fixes : brave/brave-browser#7670

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tabmodel-TabCreatorManager.java.patch [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
++import org.chromium.chrome.browser.settings.BackgroundImagesPreferences;
+ 
+ /**
+  * An interface to return a {@link TabCreator} either for regular or incognito tabs.
+@@ -90,7 +99,24 @@ public interface TabCreatorManager {
+         public final void launchNTP() {
+             try {
+                 TraceEvent.begin("TabCreator.launchNTP");
+-                launchUrl(UrlConstants.NTP_URL, TabLaunchType.FROM_CHROME_UI);
++                SharedPreferences mSharedPreferences = ContextUtils.getAppSharedPreferences();
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-tabmodel-TabCreatorManager.java.patch]: this is definitely a problem, this code should be encapsulated somewhere so we can just add a single line patch and please use the full class name for it so we don't have to add any imports

File: patches/chrome-android-java-src-org-chromium-chrome-browser-suggestions-tile-SuggestionsTileView.java.patch [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ import org.chromium.chrome.browser.suggestions.SiteSuggestion;
+ import org.chromium.chrome.browser.widget.tile.TileWithTextView;
++import org.chromium.chrome.browser.widget.tile.BraveTileWithTextView;
+ 
+ /**
+  * The view for a site suggestion tile. Displays the title of the site beneath a large icon. If a
+  * large icon isn't available, displays a rounded rectangle with a single letter in its place.
+  */
+-public class SuggestionsTileView extends TileWithTextView {
++public class SuggestionsTileView extends BraveTileWithTextView {
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-suggestions-tile-SuggestionsTileView.java.patch]: please use the full classname here to avoid the import

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ import org.chromium.ui.base.WindowAndroid;
++import org.chromium.chrome.browser.ntp.sponsored.SponsoredTab;
+ 
+ /**
+  * Implementation of the interface {@link Tab}. Contains and manages a {@link ContentView}.
+@@ -224,6 +225,8 @@ public class TabImpl implements Tab {
+ 
+     private final UserDataHost mUserDataHost = new UserDataHost();
+ 
++    private SponsoredTab sponsoredTab;
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: please use the full classname here to avoid the import

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
++
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: never add whitespace in patches

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     /**
+      * Creates an instance of a {@link TabImpl}.
+      *
+@@ -275,6 +278,9 @@ public class TabImpl implements Tab {
+                 updateInteractableState();
+             }
+         };
++
++        if(sponsoredTab == null)
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: how could this ever not be null in the initializer? In any case this should be a single line

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
++            sponsoredTab = new SponsoredTab();
+     }
+ 
+     @Override
+@@ -1564,6 +1570,10 @@ public class TabImpl implements Tab {
+         }
+     }
+ 
++    public SponsoredTab getSponsoredTab() {
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: where does this get called from?

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tabmodel-TabCreatorManager.java.patch (hunk 1 above)
Comment (bridiver) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tabmodel-TabCreatorManager.java.patch]: actually this doesn't seem like the right place for this anyway. I think we should create `BraveTabCreator` as a subclass of `ChromeTabCreator` and for now just patch in `ChromeTabbedActivity.java` and `CustomTabActivityTabFactory.java`, but we'll come back later and do it a different way with asm

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch [hunk 7]
```diff
@@ -0,0 +1,31 @@
+diff --git a/chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java b/chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java
+index d1ca3eb54eaa4988d2d347804152eae9d2381e00..c015ff58dd15669b3e6f26f8c95e57d5997dc3f8 100644
+--- a/chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java
++++ b/chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java
+@@ -99,7 +99,7 @@ public class NewTabPageView extends HistoryNavigationLayout {
+ 
+         // Don't attach now, the recyclerView itself will determine when to do it.
+         mNewTabPageLayout = (NewTabPageLayout) LayoutInflater.from(getContext())
+-                                    .inflate(R.layout.new_tab_page_layout, mRecyclerView, false);
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch]: we should just over `new_tab_page_layout` with the contents of `brave_new_tab_page_layout`. Create new_tab_page_layout.xml in brave/android/java/res/layout and it will replace the chromium version of the file

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+         mRecyclerView.setAdapter(newTabPageAdapter);
+         mRecyclerView.getLinearLayoutManager().scrollToPosition(scrollPosition);
+ 
+-        mRecyclerViewResizer = ViewResizer.createAndAttach(mRecyclerView, mUiConfig,
+-                mRecyclerView.getResources().getDimensionPixelSize(
+-                        R.dimen.content_suggestions_card_modern_margin),
+-                mRecyclerView.getResources().getDimensionPixelSize(
+-                        R.dimen.ntp_wide_card_lateral_margins));
++        if (false)
++            mRecyclerViewResizer = ViewResizer.createAndAttach(mRecyclerView, mUiConfig,
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch]: don't change the indentation here

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tabmodel-TabCreatorManager.java.patch (hunk 1 above)
Comment (deeppandya) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tabmodel-TabCreatorManager.java.patch]: @bridiver good point. i will update it.

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch (hunk 5 above)
Comment (deeppandya) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: @bridiver it's a check for specific edge case where we refresh the tab page.

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+import org.chromium.chrome.browser.ChromeFeatureList;
+import org.chromium.chrome.browser.BraveFeatureList;
+import org.chromium.chrome.browser.BraveRewardsNativeWorker;
+import org.chromium.chrome.browser.BraveRewardsPanelPopup;
+import org.chromium.chrome.browser.preferences.BravePrefServiceBridge;
+
+import static org.chromium.chrome.browser.util.ViewUtils.dpToPx;
+
+public class NTPUtil {
+	public static void openImageCredit(String url) {
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: indentation is wrong here

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch (hunk 6 above)
Comment (deeppandya) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: @bridiver we use it in BraveNewTabPageView and RewardsBottomSheetDialogFragment to get values from SponsoredTab object.

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+import org.chromium.chrome.browser.ChromeActivity;
+import org.chromium.chrome.browser.ChromeFeatureList;
+import org.chromium.chrome.browser.BraveFeatureList;
+import org.chromium.chrome.browser.BraveRewardsNativeWorker;
+import org.chromium.chrome.browser.BraveRewardsPanelPopup;
+import org.chromium.chrome.browser.preferences.BravePrefServiceBridge;
+
+import static org.chromium.chrome.browser.util.ViewUtils.dpToPx;
+
+public class NTPUtil {
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: what are these methods? Where do they get called?

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch (hunk 7 above)
Comment (deeppandya) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch]: @bridiver i did the exact same thing in previous commits but as it's more like a different UI for newtabpage. updated it this way.

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch (hunk 6 above)
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: why does sponsoredTab have to be part of TabImpl at all? It doesn't appear to need it for anything

Comment (bridiver) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: seems like it should probably be a TabHelper instead?

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java [hunk 11]
```diff
@@ -5,17 +5,97 @@
 
 package org.chromium.chrome.browser.ntp;
 
+import android.os.Bundle;
 import android.content.Context;
 import android.util.AttributeSet;
 import android.view.ViewGroup;
 import android.widget.ImageView;
 import android.widget.TextView;
+import android.content.SharedPreferences;
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java]: there are a lot of imports added here that doesn't appear to be used in this file

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch (hunk 6 above)
Comment (deeppandya) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: @bridiver we need to persist images across the tab. So i added a model class SponsoredTab as it has things only for NTP SI.

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    public static NTPImage getNTPImage() {
+        SharedPreferences mSharedPreferences = ContextUtils.getAppSharedPreferences();
+
+        if (mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_SPONSORED_IMAGES, true)
+            && SponsoredImageUtil.getSponsoredImages().size() > 0
+            && mSharedPreferences.getInt(BackgroundImagesPreferences.PREF_APP_OPEN_COUNT, 0) == 1
+            && SponsoredImageUtil.getTabIndex() == 2
+            && BraveAdsNativeHelper.nativeIsLocaleValid(Profile.getLastUsedProfile())
+            && !BravePrefServiceBridge.getInstance().getSafetynetCheckFailed()
+            && ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_REWARDS)) {
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: I don't think this is right, sponsored images are not dependent on brave rewards afaik

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+            SponsoredImage sponsoredImage = SponsoredImageUtil.getSponsoredImage();
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: indentation is wrong here, please check the entire PR for consistent 4 spaces indentation https://source.android.com/setup/contribute/code-style#use-spaces-for-indentation

Comment (bridiver) [reply android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: also here is a link for the chromium style guide where it differs from above https://chromium.googlesource.com/chromium/src/+/master/styleguide/java/java.md

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch (hunk 7 above)
Comment (bridiver) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch]: I don't understand, we are completely replacing new_tab_page_layout, right? In that case we should use the same name file so we don't have to patch here

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageLayout.java.patch [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+     private static final String TAG = "NewTabPageLayout";
+     // Used to signify the cached resource value is unset.
+     private static final int UNSET_RESOURCE_FLAG = -1;
+@@ -415,8 +416,9 @@ public class NewTabPageLayout extends LinearLayout implements TileGroup.Observer
+                 1f);
+     }
+ 
+-    private void insertSiteSectionView() {
+-        mSiteSectionView = SiteSection.inflateSiteSection(this);
++    public void insertSiteSectionView() {
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageLayout.java.patch]: why is this changed to public?

Comment (bridiver) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageLayout.java.patch]: it actually makes sense to me to change this to public and then override in `BraveNewTabPageLayout`, but instead it's changed to public and then code is inserted into it? I think what you actually want to do here is just change this to public and add this code before calling `super. insertSiteSectionView()`?

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch (hunk 5 above)
Comment (bridiver) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: but this is in the initializer for the class

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java [hunk 15]
```diff
@@ -127,4 +278,220 @@ else if (seconds > 60) {
...
+                    mNewTabPageLayout.setBackgroundResource(0);
+                    if (imageDrawable != null && imageDrawable.getBitmap() != null && !imageDrawable.getBitmap().isRecycled()) {
+                        imageDrawable.getBitmap().recycle();
+                    }
+                }
+            }
+        }
+    };
+
+    private NewTabListener newTabListener = new NewTabListener() {
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java]: is this supposed to be `NewTabPageListener`? Because `NewTabListener` doesn't seem to apply

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+                    }
+                } else if(BraveAdsNativeHelper.nativeIsLocaleValid(Profile.getLastUsedProfile())) {
+                    if (ntpImage instanceof SponsoredImage) {
+                        return SponsoredImageUtil.BR_ON_ADS_OFF ;
+                    } else {
+                        return SponsoredImageUtil.BR_ON_ADS_OFF_BG_IMAGE;
+                    }
+                }
+            } else {
+                if (ntpImage instanceof SponsoredImage && !mBraveRewardsNativeWorker.IsCreateWalletInProcess()) {
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: why do we care if the wallet creation is in process?

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+            && SponsoredImageUtil.getTabIndex() == 2
+            && BraveAdsNativeHelper.nativeIsLocaleValid(Profile.getLastUsedProfile())
+            && !BravePrefServiceBridge.getInstance().getSafetynetCheckFailed()
+            && ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_REWARDS)) {
+            SponsoredImage sponsoredImage = SponsoredImageUtil.getSponsoredImage();
+            SponsoredImageUtil.incrementTabIndex(3);
+            return sponsoredImage;
+        }
+
+        if (mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_SPONSORED_IMAGES, true)
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: there are way too many duplicated conditions here, the common conditions should be combined and then use a nested if

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+            && SponsoredImageUtil.getSponsoredImages().size() > 0
+            && SponsoredImageUtil.getTabIndex() != 1
+            && SponsoredImageUtil.getTabIndex() % 4 == 0
+            && BraveAdsNativeHelper.nativeIsLocaleValid(Profile.getLastUsedProfile())
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: these checks don't seem relevant

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    public static NTPImage getNTPImage() {
+        SharedPreferences mSharedPreferences = ContextUtils.getAppSharedPreferences();
+
+        if (mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_SPONSORED_IMAGES, true)
+            && SponsoredImageUtil.getSponsoredImages().size() > 0
+            && mSharedPreferences.getInt(BackgroundImagesPreferences.PREF_APP_OPEN_COUNT, 0) == 1
+            && SponsoredImageUtil.getTabIndex() == 2
+            && BraveAdsNativeHelper.nativeIsLocaleValid(Profile.getLastUsedProfile())
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: this check doesn't seem relevant

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageLayout.java.patch (hunk 14 above)
Comment (deeppandya) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageLayout.java.patch]: @bridiver i have tried the same thing and that's why we have public here. in this method, we add layout dynamically and we can just add layout once so instaed of creating new function, i thought of updating the same function.

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java [hunk 20]
```diff
@@ -65,7 +207,16 @@ private void updateBraveStats() {
 
         mAdsBlockedCountTextView.setText(getBraveStatsStringFormNumber(adsBlockedCount));
         mHttpsUpgradesCountTextView.setText(getBraveStatsStringFormNumber(httpsUpgradesCount));
-        mEstTimeSavedTextView.setText(getBraveStatsStringFromTime(estimatedMillisecondsSaved / 1000));
+        mEstTimeSavedCountTextView.setText(getBraveStatsStringFromTime(estimatedMillisecondsSaved / 1000));
+
+        if(mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_BACKGROUND_IMAGES, true) 
+            && (Build.VERSION.SDK_INT > Build.VERSION_CODES.M || (Build.VERSION.SDK_INT <= Build.VERSION_CODES.M && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && !sponsoredTab.isMoreTabs()))) {
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java]: this logic seems to be repeated several times and should probably be in a method

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java (hunk 15 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java]: i agree. i should change the name.

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java (hunk 16 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: from bottom banner, user can enable rewards and according to current implementation, it takes some time to create wallet and enable rewards. and if you open another tab right after enabling rewards from the bottom banner, it shows the banner again as the state is disabled. so this check is to see whether the wallet creation is in progress, if it is then we don't show the banner again

File: patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch (hunk 6 above)
Comment (bridiver) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-tab-TabImpl.java.patch]: TabHelpers exist to attach other data to Tabs. See TaskTabHelper for an example

File: android/java/org/chromium/chrome/browser/ntp/sponsored/SponsoredTab.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+    public SponsoredTab() {
+        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+        if(chromeTabbedActivity != null) {
+            TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+            isMoreTabs = tabModel.getCount() > SponsoredImageUtil.MAX_TABS ? true : false;
+        }
+
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M || (Build.VERSION.SDK_INT <= Build.VERSION_CODES.M && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && !isMoreTabs)){
+            ntpImage = NTPUtil.getNTPImage();
+            tabIndex = SponsoredImageUtil.getTabIndex();
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/ntp/sponsored/SponsoredTab.java]: I'm confused about what this is

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch (hunk 7 above)
Comment (deeppandya) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-ntp-NewTabPageView.java.patch]: @bridiver it created a patch like https://github.com/brave/brave-core/blob/e376bb777e6339e4e8210d95db313ec3f159ec1b/patches/chrome-android-java-res-layout-new_tab_page_layout.xml.patch so instead of adding patch like above i thought of creating a separate layout and just update it in java class.

File: patches/chrome-android-java-src-org-chromium-chrome-browser-customtabs-content-CustomTabActivityTabFactory.java.patch [hunk 22]
```diff
@@ -0,0 +1,31 @@
+diff --git a/chrome/android/java/src/org/chromium/chrome/browser/customtabs/content/CustomTabActivityTabFactory.java b/chrome/android/java/src/org/chromium/chrome/browser/customtabs/content/CustomTabActivityTabFactory.java
+index 157946aad58b6efcba212da92effb3dc9e7b30d2..db7638335f72755418c6437894742836b73ca219 100644
+--- a/chrome/android/java/src/org/chromium/chrome/browser/customtabs/content/CustomTabActivityTabFactory.java
++++ b/chrome/android/java/src/org/chromium/chrome/browser/customtabs/content/CustomTabActivityTabFactory.java
+@@ -32,7 +32,7 @@ import javax.inject.Inject;
+ import dagger.Lazy;
+ 
+ /**
+- * Creates {@link Tab}, {@link TabModelSelector}, and {@link ChromeTabCreator}s in the context of a
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-customtabs-content-CustomTabActivityTabFactory.java.patch]: never patch comments

File: patches/chrome-android-java-src-org-chromium-chrome-browser-customtabs-content-CustomTabActivityTabFactory.java.patch [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
++ * Creates {@link Tab}, {@link TabModelSelector}, and {@link org.chromium.chrome.browser.tabmodel.BraveTabCreator}s in the context of a
+  * Custom Tab activity.
+  */
+ @ActivityScope
+@@ -78,13 +78,13 @@ public class CustomTabActivityTabFactory {
+         return mTabModelSelector;
+     }
+ 
+-    /** Creates a {@link ChromeTabCreator}s for the custom tab. */
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-customtabs-content-CustomTabActivityTabFactory.java.patch]: these methods don't have to change because BraveTabCreator is a subclass of ChromeTabCreator and you can cast if you need anything in `BraveTabCreator` that is not in `ChromeTabCreator`

File: patches/chrome-android-java-src-org-chromium-chrome-browser-customtabs-CustomTabActivity.java.patch [hunk 24]
```diff
@@ -0,0 +1,13 @@
+diff --git a/chrome/android/java/src/org/chromium/chrome/browser/customtabs/CustomTabActivity.java b/chrome/android/java/src/org/chromium/chrome/browser/customtabs/CustomTabActivity.java
+index 31bfebf2e1faf4620a3f943f5212918fb91b6ec1..d679f39d3d2eda0476d4b248ef6063f32b1ab1b1 100644
+--- a/chrome/android/java/src/org/chromium/chrome/browser/customtabs/CustomTabActivity.java
++++ b/chrome/android/java/src/org/chromium/chrome/browser/customtabs/CustomTabActivity.java
+@@ -195,7 +195,7 @@ public class CustomTabActivity extends BaseCustomTabActivity<CustomTabActivityCo
+     }
+ 
+     @Override
+-    protected Pair<ChromeTabCreator, ChromeTabCreator> createTabCreators() {
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-customtabs-CustomTabActivity.java.patch]: same as below, no need to patch this

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ChromeTabbedActivity.java.patch [hunk 25]
```diff
@@ -11,3 +11,47 @@ index be8466ad87791880f1f3c4ea61db11c466eb3356..9d977486ffffda3c6545d974c1f4e8a2
      /**
       * The results of a system back press action.
       */
+@@ -965,13 +965,13 @@ public class ChromeTabbedActivity extends ChromeActivity implements ScreenshotMo
+     }
+ 
+     @Override
+-    public ChromeTabCreator getTabCreator(boolean incognito) {
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-ChromeTabbedActivity.java.patch]: these patches are also not needed

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ChromeTabbedActivity.java.patch [hunk 26]
```diff
@@ -11,3 +11,47 @@ index be8466ad87791880f1f3c4ea61db11c466eb3356..9d977486ffffda3c6545d974c1f4e8a2
...
+     }
+ 
+     @Override
+-    protected Pair<ChromeTabCreator, ChromeTabCreator> createTabCreators() {
++    protected Pair<org.chromium.chrome.browser.tabmodel.BraveTabCreator, org.chromium.chrome.browser.tabmodel.BraveTabCreator> createTabCreators() {
+         Supplier<TabDelegateFactory> tabDelegateFactorySupplier = () -> {
+             return new TabbedModeTabDelegateFactory(this);
+         };
+-        return Pair.create(new ChromeTabCreator(this, getWindowAndroid(), getStartupTabPreloader(),
++        return Pair.create(new org.chromium.chrome.browser.tabmodel.BraveTabCreator(this, getWindowAndroid(), getStartupTabPreloader(),
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-ChromeTabbedActivity.java.patch]: this patch is correct to change the class that is actually created, but the patch for the method signature is not needed

File: patches/chrome-android-java-res-layout-new_tab_page_layout.xml.patch [hunk 27]
```diff
@@ -11,9 +11,175 @@ index 522b6b1bfb35b83e758995459e00847bfad1ada8..9ce74212d300394a85bcf5ca7d12167e
      xmlns:android="http://schemas.android.com/apk/res/android"
      xmlns:app="http://schemas.android.com/apk/res-auto"
      xmlns:tools="http://schemas.android.com/tools"
-@@ -71,4 +71,4 @@
-         android:layout_height="0dp"
-         android:layout_weight="1"
-         android:visibility="gone" />
+     android:id="@+id/ntp_content"
```
Comment (bridiver) [inline patches/chrome-android-java-res-layout-new_tab_page_layout.xml.patch]: Looks like you accidentally applied this as a patch. You can just leave it out and it will work automatically when your new_tab_page_layout.xml file is copied over

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java (hunk 11 above)
Comment (bridiver) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageView.java]: did you check these? Still looks like a ton of unused imports

File: android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java (hunk 13 above)
Comment (bridiver) [reply android/java/org/chromium/chrome/browser/ntp/sponsored/NTPUtil.java]: I thought we were going to switch to using the c++ code for this stuff? We shouldn't be duplicating it in java

**Comment by deeppandya** [comment]
Created new PR : https://github.com/brave/brave-core/pull/4807
